alf/environments/suite_carla.py

```diff
@@ -0,0 +1,1163 @@
+            pygame.quit()
+
+        return commands
+
+    def observation_spec(self):
+        """Get the observation spec.
+
+        Returns:
+            nested TensorSpec:
+        """
+        return self._observation_spec
+
+    def observation_desc(self):
+        """Get the description about the observation.
+
+        Returns:
+            nested str: each str corresponds to one TensorSpec from
+            ``observatin_spec()``.
+        """
+        return self._observation_desc
+
+    def action_spec(self):
+        """Get the action spec.
+
+        Returns:
+            nested BoundedTensorSpec:
+        """
+        return alf.BoundedTensorSpec([4],
+                                     minimum=[0., -1., 0., 0.],
+                                     maximum=[1., 1., 1., 1.])
+
+    def action_desc(self):
+        """Get the description about the action.
+
+        Returns:
+            nested str: each str corresponds to one TensorSpec from
+            ``action_spec()``.
+        """
+        return (
+            "4-D vector of [throttle, steer, brake, reverse], where "
+            "throttle is in [0.0, 1.0], steer is in [-1.0, 1.0], brake is in "
+            "[0.0, 1.0], and reverse is interpreted as a boolean value "
+            "with values greater than 0.5 corrsponding to True.")
+
+    def _get_goal(self):
+        trans = self._actor.get_transform()
+        loc = trans.location
+        yaw = math.radians(trans.rotation.yaw)
+
+        loc = np.array([loc.x, loc.y, loc.z])
+        cos, sin = np.cos(yaw), np.sin(yaw)
+        rot = np.array([[cos, sin, 0.], [-sin, cos, 0.], [0., 0., 1.]])
+        return np.matmul(rot, self._goal_location - loc).astype(np.float32)
+
+    def get_current_time_step(self, current_frame):
+        """Get the current time step for the player.
+
+        Args:
+            current_frame (int): current simulation frame no.
+        Returns:
+            TimeStep: all elements are nd.ndarray or np.number.
+        """
+        obs = OrderedDict()
+        for sensor_name, sensor in self._observation_sensors.items():
+            obs[sensor_name] = sensor.get_current_observation(current_frame)
+        obs['goal'] = self._get_goal()
+        v = self._actor.get_velocity()
+        obs['speed'] = np.float32(
+            np.float32(v.x * v.x + v.y * v.y + v.z * v.z))
+        self._current_distance = np.linalg.norm(obs['goal'])
+        reward = 0.
+        if self._fail_frame is None:
+            step_type = ds.StepType.FIRST
+            max_frames = math.ceil(
+                self._current_distance / self._min_speed / self._delta_seconds)
+            self._fail_frame = current_frame + max_frames
+        elif (self._current_distance < self._success_distance_thresh
+              and self._actor.get_velocity() == carla.Location()):
+            # TODO: include waypoint orientation as success critiria
+            self._done = True
+            step_type = ds.StepType.LAST
```

Comment:
In this case the discount should be 0? Otherwise the step_type should still be MID. For the former case, need to set self._fail_frame = current_frame.

Reply:
Added discount=0. I don't understand why need to set self._fail_frame=current_frame